feat (DOCS-CODE-028 | DOCS-CODE-033): add iOS and macOS SwiftUI canonical tutorials with MSAL (#1)
Azure-Samples/ms-identity-docs-code-swift: src/msalexample-ios-macos/README.md

<!-- Keeping yaml frontmatter commented out for now
---
# Metadata required by https://docs.microsoft.com/samples/browse/
# Metadata properties: https://review.docs.microsoft.com/help/contribute/samples/process/onboarding?branch=main#add-metadata-to-readme
languages:
- swift 
page_type: sample
name: "SwiftUI Multiplatform app that makes a request to the Graph API after signing in the user"
description: "This Swift 5.5 Multiplatform SwiftUI iOS and macOS app signs in the user and then makes a request to Microsoft Graph for the user's profile data."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-app-swift-multiplatform
---
-->

> # Swift | Multiplatform SwiftUI iOS and macOS | user sign-in, protected web API access (Microsoft Graph) | Microsoft identity platform

Reviewer: Should we call this "Native," or maybe "Desktop and mobile?"

Xplat is fine, but thinking maybe we stick with our naming scheme.

Author: The current one is the one that feels more idiomatic. But open to get rid of "Multiplatform" and just keep "SwiftUI". 

Open to go for another name including "Native" or "Desktop and mobile".

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This Swift 5.5 Multiplatform SwiftUI iOS and macOS app authenticates a user and then makes a request to the Graph API as the authenticated user. The response to the request is presented to the user.

![A screenshot of a Swift 5.5 Multiplatform SwiftUI iOS app displaying a response from Microsoft Graph.](./ios-signout-app.png)

> ![A screenshot of a Swift 5.5 Multiplatform SwiftUI macOS app displaying a response from Microsoft Graph.](./macos-signout-app.png)

Reviewer: ```suggestion
![A screenshot of a Swift 5.5 Multiplatform SwiftUI iOS app displaying a response from Microsoft Graph.](./ios-signout-app.png)

![A screenshot of a Swift 5.5 Multiplatform SwiftUI macOS app displaying a response from Microsoft Graph.](./macos-signout-app.png)
```

## Prerequisites

- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
- Xcode 13.2.1

> - Swift 5.5

Reviewer: Is it _only_ these versions, or can we do "+" after the versions to indicate "above?"

Author: I think it is safe for now to add `+` over here.

```suggestion
- Xcode 13.2.1
- Swift 5+
```

|--------------------------------:|:--------------------------------------------------------------------|:--------------------------------------------------------------------------------|
| **Name**                        | `active-directory-swift-ios-macos-authorization-code-grant-flow`    | Suggested value for this sample. <br/> You can change the app name at any time. |
| **Supported account types**     | **Accounts in this organizational directory only (Single tenant)**  | Suggested value for this sample.                                                |
| **Platform type**               | **iOS / macOS **                                                    | Required value for this sample                                                  |
| **Redirect URIs**               | `msauth.com.contoso.msalExample-macOS://auth` and `msauth.com.contoso.msalExample-iOS://auth` | Required value for this sample                    |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Open the project in Xcode 

> Next, open the _msalexamples-ios-macos.xcodeproj_ project in Xcode.

Reviewer: :question: **Side question:** How was this project initialized? Is there an analog to `dotnet new $APP_TYPE` that you used for firing up a blank project that you then added the MSAL stuff to?

Asking for a friend who's never used Xcode. :flushed:

Author: just pure Xcode GUI.  

`Create a new Xcode project` -> `Multiplatform` -> `App`

or

`Create a new Xcode project` -> `Other` -> `Empty`

### 3. Update code sample in MSALAuthentication.swift_ with app registration values

Finally, set the following values in _Shared/MSALAuthentication.swift_.

```swift
// 'Application (client) ID' of app registration in Azure portal - this value is a GUID
private static let kClientId = ""

// 'Tenant ID' of your Azure AD instance - this value is a GUID
private static let kTenantId = ""
```

## Run the application

> Select My Mac or any iOS simulator and then press :arrow_forward: 

Reviewer: ```suggestion
Select **My Mac** or any iOS simulator and then select :arrow_forward: 
```

> The appliction will open allowing you to click the **Sign In** button to use the authentication flow.

Reviewer: ```suggestion
The application will open, allowing you to click the **Sign In** button to initiate the authentication flow.
```

![A screenshot of a Swift 5.5 Multiplatform SwiftUI iOS app guiding the user to click the "Sign In" button.](./ios-signin-app.png)

> ![A screenshot of a Swift 5.5 Multiplatform SwiftUI macOS app guiding the user to click the "Sign In" button.](./macos-signin-app.png)

Reviewer: ```suggestion
![A screenshot of a Swift 5.5 Multiplatform SwiftUI iOS app guiding the user to click the "Sign In" button.](./ios-signin-app.png)

![A screenshot of a Swift 5.5 Multiplatform SwiftUI macOS app guiding the user to click the "Sign In" button.](./macos-signin-app.png)
```

## About the code

This Swift 5.5 Multiplatform iOS and macOS app presents a button that initiates an authentication flow using the Microsoft Authentication Library for iOS and macOS (MSAL). The user completes this flow in their default web browser or natively in iOS. Upon successful authentication, an HTTP GET request to the Microsoft Graph /me endpoint is issued with the user's access token in the HTTP header. The response from the GET request is then displayed to the user.

> This project use Swift Package Manager dependencies to reference `MSAL` from [AzureAD/microsoft-authentication-library-for-objc](https://github.com/AzureAD/microsoft-authentication-library-for-objc). For more information, please refer to using [Swift Packages](https://github.com/AzureAD/microsoft-authentication-library-for-objc/tree/3bc25ad3c38c0f0044e3fc624a841ac4789478c0#using-swift-packages)

Reviewer: ```suggestion
This project uses Swift Package Manager dependencies to reference `MSAL` from [AzureAD/microsoft-authentication-library-for-objc](https://github.com/AzureAD/microsoft-authentication-library-for-objc). For more information, please refer to using [Swift Packages](https://github.com/AzureAD/microsoft-authentication-library-for-objc/tree/3bc25ad3c38c0f0044e3fc624a841ac4789478c0#using-swift-packages)
```

| **Supported account types**     | **Accounts in this organizational directory only (Single tenant)**  | Suggested value for this sample.                                                |
| **Platform type**               | **iOS / macOS **                                                    | Required value for this sample                                                  |
| **Redirect URIs**               | `msauth.com.microsoft.msalExample-macOS://auth` and `msauth.com.microsoft.msalExample-iOS://auth` | Required value for this sample                    |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

### 2. Open the project in Xcode 

Next, open the _msalexamples-ios-macos.xcodeproj_ project in Xcode.

### 3. Update code sample in MSALAuthentication.swift_ with app registration values

Finally, set the following values in _Shared/MSALAuthentication.swift_.

> ```csharp

Reviewer 2: ~~~suggestion
```swift
~~~

// 'Application (client) ID' of app registration in Azure portal - this value is a GUID
private static let kClientId = "APPLICATION_(CLIENT)_ID"

// 'Tenant ID' of your Azure AD instance - this value is a GUID
private static let kTenantId = "TENANT_ID"
```

## Run the application

> iOS
> 
> Select any iOS simulator and then press :arrow_forward: 
> 
> macOS
> 
> Select My Mac and then press :arrow_forward: 

Reviewer 2: Suggestion to make it even tighter.

```suggestion
Select My Mac or any iOS simulator and then press :arrow_forward: 
```
